### ISSUE TITLE
[turborepo] optional ENV vars don't invalidate cache when removed in later run

### ISSUE THREAD

**Issue by CITguy** [vercel/turbo#4645]
### What version of Turborepo are you using?

~1.8.3~ 1.9.3

### What package manager are you using / does the bug impact?

npm

### What operating system are you using?

Mac

### Describe the Bug

The turbo cache is invalidated when _adding_ an **optional** ENV var, but not when _removing_.

1. `turbo run build` (cache MISS)
2. `buildtarget=prod turbo run build` (cache MISS)
3. `turbo run build` (cache HIT)

### Expected Behavior

1. `turbo run build` (cache MISS)
2. `buildtarget=prod turbo run build` (cache MISS)
3. `turbo run build` (cache MISS)

### To Reproduce

```javascript
{
  "globalEnv": [ "buildtarget" ],
  "pipeline": {
    "build": {
      "outputs": [ "dist/**" ]
    }
  }
}
```

**Comment by CITguy** [comment]
To clarify, if `buildtarget` is not present, it defaults to `dev`, so the following commands are equivalent:

- `turbo run build`
- `buildtarget=dev turbo run build`

However, the majority of the time, devs want the default, so using the shorthand is preferable.  

Additionally, the same logic is used for optional "boolean" vars like `buildminified=true` (defaults to `buildminified=false`).  During normal development, minification is undesirable, so we only enable it when we need to publish a new release.

**Comment by CITguy** [comment]
TL;DR `config.globalEnv` and `config.pipeline.<task>.env` values should ALWAYS be included in the caching algorithm, regardless of their ENV presence (i.e., **null/missing ENV vars are _significant_ when busting the cache**).

Assuming that I have the following turbo config...

```javascript
{
  "globalEnv": [ 
    "buildtarget",
    "buildminified"
  ],
  "pipeline": { 
    "build": {}
  }
}
```

From what I can gather, if an ENV var is _not_ present, it's not being included in determining the cache key in the caching algorithm.

I'd expect that ALL of the following commands would result in a `cache MISS`, if run in a random order.

command | cache.key
----- | -----
`turbo run build` | `buildtarget=null\|buildminified=null`
`buildtarget=prod turbo run build` | `buildtarget=prod\|buildminified=null`
`buildtarget=dev turbo run build` | `buildtarget=dev\|buildminified=null`
`buildtarget=prod buildminified=true turbo run build` | `buildtarget=prod\|buildminified=true`
`buildtarget=dev buildminified=true turbo run build` | `buildtarget=dev\|buildminified=true`
`buildminified=true turbo run build` | `buildtarget=null\|buildminified=true`

**Comment by nathanhammond** [comment]
Hey @CITguy! Thank you for the report! We wanted to take the time and make sure that we had the correct behavior here.

I added a few tests into our environment variable test suite which correspond to your examples _identically_:
<img width="631" alt="Added tests." src="https://user-images.githubusercontent.com/20542/234537119-24c7d26c-06f9-46bd-bd4b-2797047f5cfc.png">

That shows that we produce a cache key for environment variables that corresponds to the provided input every time.

**Comment by nathanhammond** [comment]
It took me a while to identify _why_ you were seeing unexpected behavior, but it's because your reproduction steps are doing something that surprised you:

Your run is this:
```
1. turbo run build (cache MISS)
2. buildtarget=prod turbo run build (cache MISS)
3. turbo run build (cache HIT)
```

What is happening on line 3 is that it is hitting a cache from run 1, not from run 2. Cache hits are not time-ordered; they're content-hashed. Since we detect that run 1 & run 3 are _identical_ runs, we match the cache from run 1 on run 3.

You can see this happening in more detail if you pass `--dry-json` and look at the inputs to the hashes and the hash values. You'll see that run 3 returns the same hash as run 1.

***

One last thing, since `turbo` doesn't know about your application logic, these two commands are _not_ identical as far as `turbo` is concerned:

```
turbo run build
buildtarget=dev turbo run build
```

This may be the default in application logic, but `turbo` is unable to know that it can treat `buildtarget=` and `buildtarget=dev` as identical. I'd instead recommend that you remove `buildtarget=dev` as an actual input option, standardizing on an unspecified `buildtarget` meaning that you're building for `dev`. Because right now you actually have _three_ `buildtarget` input configurations that `turbo` will differentiate between:

1. `buildtarget=`
2. `buildtarget=dev`
3. `buildtarget=prod`
